### ISSUE TITLE
unexpected behavior of defaultTags

### ISSUE THREAD

**Issue by filip-zyzniewski** [pulumi/pulumi-aws#1655]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

The AWS provider ignores changes to defaultTags after a resource has been created unless other attributes are being changed

### Steps to reproduce

1. Specify some defaultTags and create a bucket. The bucket has the tags as expected:

```
~/tmp/wtf/defaultTags$ cat index.ts
import * as aws from "@pulumi/aws";

const bucket = new aws.s3.Bucket("bucket-with-default-tags", {bucket: "bucket-with-default-tags"});
~/tmp/wtf/defaultTags$ cat Pulumi.dev.yaml
encryptionsalt: v1:ggyom6IhdG8=:v1:sJqljlbglc8JJgpe:jYzdmym/+StNPwPbn+7uIpAIL+fKLA==
config:
  aws:region: eu-west-1
  aws:defaultTags:
    tags:
      Tag1: value1
      Tag2: value2
~/tmp/wtf/defaultTags$ pulumi up --diff
Previewing update (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    + aws:s3/bucket:Bucket: (create)
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::04da6b54-80e4-46f7-96ec-b56ff0331ba9]
        acl         : "private"
        bucket      : "bucket-with-default-tags"
        forceDestroy: false

import * as aws from "@pulumi/aws";
Resources:
    + 1 to create
    1 unchanged
Do you want to perform this update? yes
Updating (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    + aws:s3/bucket:Bucket: (create)
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::8510e26a-009d-460f-a5e1-4f5e95a6af22]
        acl         : "private"
        bucket      : "bucket-with-default-tags"
        forceDestroy: false
Resources:
    + 1 created
    1 unchanged

Duration: 11s
~/tmp/wtf/defaultTags$ aws s3api get-bucket-tagging --bucket bucket-with-default-tags
{
    "TagSet": [
        {
            "Key": "Tag1",
            "Value": "value1"
        },
        {
            "Key": "Tag2",
            "Value": "value2"
        }
    ]
}
```

2. Add some new tags to the defaultTags and attempt an update, which shows no diff and results in no changes:

```
~/tmp/wtf/defaultTags$ echo '      Tag3: value3' >> Pulumi.dev.yaml
~/tmp/wtf/defaultTags$ echo '      Tag4: value4' >> Pulumi.dev.yaml
~/tmp/wtf/defaultTags$ cat Pulumi.dev.yaml
encryptionsalt: v1:ggyom6IhdG8=:v1:sJqljlbglc8JJgpe:jYzdmym/+StNPwPbn+7uIpAIL+fKLA==
config:
  aws:region: eu-west-1
  aws:defaultTags:
    tags:
      Tag1: value1
      Tag2: value2
      Tag3: value3
      Tag4: value4
~/tmp/wtf/defaultTags$ pulumi up --diff
Previewing update (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
Resources:
    2 unchanged
Do you want to perform this update? yes
Updating (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
Resources:
    2 unchanged

Duration: 3s
~/tmp/wtf/defaultTags$ aws s3api get-bucket-tagging --bucket bucket-with-default-tags
{
    "TagSet": [
        {
            "Key": "Tag1",
            "Value": "value1"
        },
        {
            "Key": "Tag2",
            "Value": "value2"
        }
    ]
}
```

3. Change another property of the bucket and observe how the new tags are also added to the bucket (but are not being shown in the preview):

```
~/tmp/wtf/defaultTags$ vim index.ts
~/tmp/wtf/defaultTags$ cat index.ts
import * as aws from "@pulumi/aws";

const bucket = new aws.s3.Bucket("bucket-with-default-tags", {
    bucket: "bucket-with-default-tags",
    acl: "authenticated-read",
});
~/tmp/wtf/defaultTags$ pulumi up --diff
Previewing update (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    ~ aws:s3/bucket:Bucket: (update)
        [id=bucket-with-default-tags]
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::8510e26a-009d-460f-a5e1-4f5e95a6af22]
      ~ acl: "private" => "authenticated-read"
Resources:
    ~ 1 to update
    1 unchanged
Do you want to perform this update? yes
Updating (dev):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::defaultTags::pulumi:pulumi:Stack::defaultTags-dev]
    ~ aws:s3/bucket:Bucket: (update)
        [id=bucket-with-default-tags]
        [urn=urn:pulumi:dev::defaultTags::aws:s3/bucket:Bucket::bucket-with-default-tags]
        [provider=urn:pulumi:dev::defaultTags::pulumi:providers:aws::default_4_23_0::8510e26a-009d-460f-a5e1-4f5e95a6af22]
      ~ acl: "private" => "authenticated-read"
        --outputs:--
      ~ tagsAll                          : {
            Tag1: "value1"
            Tag2: "value2"
          + Tag3: "value3"
          + Tag4: "value4"
        }
Resources:
    ~ 1 updated
    1 unchanged

Duration: 15s
~/tmp/wtf/defaultTags$ aws s3api get-bucket-tagging --bucket bucket-with-default-tags
{
    "TagSet": [
        {
            "Key": "Tag1",
            "Value": "value1"
        },
        {
            "Key": "Tag4",
            "Value": "value4"
        },
        {
            "Key": "Tag3",
            "Value": "value3"
        },
        {
            "Key": "Tag2",
            "Value": "value2"
        }
    ]
}
~/tmp/wtf/defaultTags$
```

Expected:

- I expected the Tag3 and Tag4 to be added to the bucket in step 2.
- I also expected the preview in step 3 to show the addition of Tag3 and Tag4 given that it has added them in the update.

Actual: 

- Tag3 and Tag4 were not added in step 2
- Addition of Tag3 and Tag4 was not shown in preview of step 3

Inspiration for this issue report: https://stackoverflow.com/a/68829181/13846505

**Comment by filip-taxamo** [comment]
@leezen any updates on this by any chance :)?

**Comment by mwcampbell** [comment]
Anyone know of a workaround for this? I already have a stack with hundreds of resources, and I want to add a common tag to all of them. I was hoping I could use `defaultTags` to do this. Is there another way that doesn't require manually tagging them all?

**Comment by filip-zyzniewski** [comment]
@mwcampbell you can use something like https://www.pulumi.com/blog/automatically-enforcing-aws-resource-tagging-policies/#automatically-applying-tags .

**Comment by adam-kiss-sg** [comment]
Any update on this? Or workaround? Almost a year later this still seems to be an issue, defaultTags don't seem to be appearing in any diff at all, and it also doesn't trigger the update of resources.

@filip-zyzniewski your alternative sounds promising, but it includes maintaining a long list of resources which are taggable (https://github.com/joeduffy/aws-tags-example/blob/master/autotag-py/taggable.py). Or is there any source of those resources? (eg: somewhere in the pulumi sdk?)

**Comment by filip-zyzniewski** [comment]
> @filip-zyzniewski your alternative sounds promising, but it includes maintaining a long list of resources which are taggable (https://github.com/joeduffy/aws-tags-example/blob/master/autotag-py/taggable.py).

The way we are approaching this now is:

```python
import pulumi
import pulumi.resource as res
import typeguard


def is_taggable(resource):
    internal_init = getattr(type(resource), "_internal_init", None)
    if not internal_init:
        return False
    tags_type = internal_init.__annotations__.get("tags", None)
    if not tags_type:
        return False
    try:
        typeguard.check_type("tags", {"": ""}, tags_type)
    except TypeError:
        return False
    return True


def register_auto_tags(**auto_tags):
    def add_tags(
        args: res.ResourceTransformationArgs,
    ) -> res.ResourceTransformationResult:
        props = args.props
        if is_taggable(args.resource):
            props = dict(props)
            props["tags"] = auto_tags | (props.get("tags") or {})
        return res.ResourceTransformationResult(props=props, opts=args.opts)

    pulumi.runtime.register_stack_transformation(add_tags)
```

**Comment by tmehlinger** [comment]
Also running into this. A fix would be much appreciated.

**Comment by NWCalvank** [comment]
This behaviour makes defaultTags effectively useless.

Is there a clear root cause of the issue? If [the Terraform provider performs as-expected](https://github.com/pulumi/pulumi-aws/issues/1655#issuecomment-938417222) then this is a Pulumi-specific issue, correct?

If anyone internal to Pulumi could point the community in the right direction here, I'm sure there might be someone (possibly me) who would be willing to help with pushing a fix.

**Comment by AaronFriel** [comment]
I'm pleased to say that the next release of the AWS provider v5 will contain a fix for this issue. Thanks all for your patience and understanding.

**Comment by jdonohoo** [comment]
Thank you for jumping on this, I was one of the enterprise customers who called out how bad this issue was impacting my organization.

**Comment by estiller** [comment]
@AaronFriel, I updated `@pulumi/aws` to the latest version, and indeed `defaultTags` were updated for most resources. However, this still doesn't work for resources created with `@pulumi/awsx`. Will you publish a new version of that package as well?

To clarify, I have resources created with `@pulumi/aws` and `@pulumi/awsx` in the same stack. The resources created with `@pulumi/aws` picked up the new default tags as expected, but those created via `@pulumi/awsx` were not updated and kept the old, bad behaviour.
